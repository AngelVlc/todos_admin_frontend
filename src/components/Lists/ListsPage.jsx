import React, { useEffect, useContext, useState } from 'react';
import { Link, useHistory } from "react-router-dom";
import { AppContext } from '../../contexts/AppContext';
import { doGet } from '../../helpers/api';

export const ListsPage = () => {
    const { auth, requestsDispatch } = useContext(AppContext)
    const [lists, setLists] = useState([]);
    let history = useHistory();

    useEffect(() => {
        const getLists = async () => {
            const res = await doGet('lists', auth.info.token, requestsDispatch)
            setLists(res);
        }
        if (auth.info) {
            getLists();
        }
    }, [auth.info, requestsDispatch]);

    return (
        <div className="container">
            <h3 className="title">LISTS</h3>
            <table className="table">
                <thead>
                    <tr>
                        <td>Name</td>
                        <td>
                            <button className="button is-small" data-testid="addNew" onClick={() => history.push('lists/new')}>
                                <span className="icon is-small">
                                    <i className="fas fa-plus"></i>
                                </span>
                            </button>
                        </td>
                    </tr>
                </thead>
                <tbody>
                    {lists.length > 0 && lists.map((list) => (
                        <tr key={list.id}>
                            <td>
                                <Link className="has-text-black" data-testid={`editList${list.id}`} to={`/lists/${list.id}/edit`}>{list.name}</Link>
                            </td>
                            <td>
                                <center>
                                    <Link className="has-text-black" data-testid={`deleteList${list.id}`} to={`/lists/${list.id}/delete`}>
                                        <span className="icon is-small">
                                            <i className="fas fa-trash-alt fa-xs"></i>
                                        </span>
                                    </Link>
                                </center>
                            </td>
                        </tr>
                    ))
                    }
                </tbody>
            </table>
        </div>
    )
}
