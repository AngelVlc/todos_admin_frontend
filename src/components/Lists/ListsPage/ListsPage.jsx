import React from 'react';
import { useHistory } from "react-router-dom";

export const ListsPage = () => {
    let history = useHistory();

    return (
        <div className="container">
            <h3 className="title">LISTS</h3>
            <button className="button" onClick={() => history.push('/')}>Home</button>
        </div>
    )
}
