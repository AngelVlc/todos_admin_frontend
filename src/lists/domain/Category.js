export class Category {
  constructor({ id, name, description }) {
    this.id = id;
    this.name = name;
    this.description = description;
  }

  static createEmpty() {
    return new Category({ id: -1, name: "", description: "" });
  }
}
