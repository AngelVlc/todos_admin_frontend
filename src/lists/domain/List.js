export class List {
  constructor({ id, name, itemsCount, items = [] }) {
    this.id = id;
    this.name = name;
    this.items = items;
    this.itemsCount = itemsCount;
  }

  static createEmpty() {
    return new List({ id: -1, name: "", itemsCount: 0 });
  }

  addNewItem(newItem) {
    newItem.id = -1;

    for (const item of this.items) {
      if (item.id <= newItem.id) {
        newItem.id = item.id - 1;
      }
    };

    this.items.push(newItem);
  }
}
