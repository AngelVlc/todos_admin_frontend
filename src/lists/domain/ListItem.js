export class ListItem {
  constructor({ id, title, description, listId }) {
    this.id = id;
    this.title = title;
    this.description = description;
    this.listId = listId;
  }

  static createEmpty(listId) {
    return new ListItem({ id: -1, listId: listId, title: "", description: "" });
  }
}
