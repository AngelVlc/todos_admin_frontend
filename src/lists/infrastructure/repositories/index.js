export * from "./ListsRepository";
export * from "./ListItemsRepository";
