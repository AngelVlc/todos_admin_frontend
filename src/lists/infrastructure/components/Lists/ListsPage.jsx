import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListsUseCase } from "../../../application/lists";

export const ListsPage = () => {
  const [lists, setLists] = useState([]);
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getLists = useCallback(async () => {
    const getListsUseCase = useCaseFactory.get(GetListsUseCase);
    const data = await getListsUseCase.execute();

    setLists(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getLists();
  }, [getLists]);

  const onNewClick = () => {
    history.push("/lists/new");
  };

  return (
    <div className="container">
      <h3 className="title">LISTS</h3>
      <nav className="breadcrumb" aria-label="breadcrumbs">
        <ul>
          <li>
            <Link to={`/`}>Home</Link>
          </li>
          <li className="is-active">
            <Link aria-current="page" to={`/lists`}>
              Lists
            </Link>
          </li>
        </ul>
      </nav>
      <table className="table">
        <thead>
          <tr>
            <td>Name</td>
            <td># Items</td>
            <td>
              <button
                className="button is-small"
                data-testid="addNew"
                onClick={() => onNewClick()}
              >
                <span className="icon is-small">
                  <i className="fas fa-plus"></i>
                </span>
              </button>
            </td>
          </tr>
        </thead>
        <tbody>
          {lists.length > 0 &&
            lists.map((list) => (
              <tr key={list.id}>
                <td>
                  <Link
                    className="has-text-black"
                    data-testid={`editList${list.id}`}
                    to={`/lists/${list.id}/edit`}
                  >
                    {list.name}
                  </Link>
                </td>
                <td>
                  <span>
                    <center>{list.itemsCount}</center>
                  </span>
                </td>
                <td>
                  <center>
                    <Link
                      className="has-text-black delete"
                      data-testid={`deleteList${list.id}`}
                      to={`/lists/${list.id}/delete`}
                    >
                    </Link>
                  </center>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};
