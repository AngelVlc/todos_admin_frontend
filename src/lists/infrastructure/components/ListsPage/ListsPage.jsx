import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListsUseCase } from "../../../application/lists";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";

export const ListsPage = () => {
  const [lists, setLists] = useState();
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getLists = useCallback(async () => {
    const getListsUseCase = useCaseFactory.get(GetListsUseCase);
    const data = await getListsUseCase.execute();

    setLists(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getLists();
  }, [getLists]);

  const onNewClick = () => {
    history.push("/lists/new");
  };

  return (
    <>
      {lists && (
        <div className="container">
          <Breadcrumb items={[{ url: "/lists", text: "Lists" }]} />
          <h3 className="title">LISTS</h3>
          <table className="table">
            <thead>
              <tr>
                <td>Name</td>
                <td># Items</td>
                <td>
                  <button
                    className="button is-small"
                    data-testid="addNew"
                    onClick={() => onNewClick()}
                  >
                    <span className="icon is-small">
                      <i className="fas fa-plus"></i>
                    </span>
                  </button>
                </td>
              </tr>
            </thead>
            <tbody>
              {lists.length > 0 &&
                lists.map((list) => (
                  <tr key={list.id}>
                    <td>
                      <Link
                        className="has-text-black"
                        data-testid={`viewList${list.id}`}
                        to={`/lists/${list.id}`}
                      >
                        {list.name}
                      </Link>
                    </td>
                    <td>
                      <span>
                        <center>{list.itemsCount}</center>
                      </span>
                    </td>
                    <td>
                      <center>
                        <Link
                          className="has-text-black delete"
                          data-testid={`deleteList${list.id}`}
                          to={`/lists/${list.id}/delete`}
                        ></Link>
                      </center>
                    </td>
                  </tr>
                ))}
            </tbody>
          </table>
        </div>
      )}
    </>
  );
};
