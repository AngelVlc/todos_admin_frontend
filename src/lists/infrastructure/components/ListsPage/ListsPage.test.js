import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { ListsPage } from "./ListsPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { List } from "../../../domain";

afterEach(cleanup);

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const history = createMemoryHistory();

const renderWithContextAndRouter = () => {
  const fakeGetListsUseCase = {
    execute: () => [
      new List({ id: 1, name: "list1", itemsCount: 4 }),
      new List({ id: 2, name: "list2", itemsCount: 6 }),
    ],
  };

  const useCaseFactory = {
    get: () => fakeGetListsUseCase,
  };

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <Router history={history}>
          <ListsPage />
        </Router>
      </AppContext.Provider>
    ),
  };
};

describe("ListPage", () => {
  it("should match the snapshot", async () => {
    let fragment;
    await act(async () => {
      const { asFragment } = renderWithContextAndRouter();
      fragment = asFragment;
    });

    expect(fragment()).toMatchSnapshot();
  });

  it("should add a new list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("addNew"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/new");
    mockHistoryPush.mockClear();
  });

  it("should view the list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("viewList2"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
    mockHistoryPush.mockClear();
  });

  it("should delete the list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("deleteList2"));
    });

    expect(history.location.pathname).toBe("/lists/2/delete");
  });
});
