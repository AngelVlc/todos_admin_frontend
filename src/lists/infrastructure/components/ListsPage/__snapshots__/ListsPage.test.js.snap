// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists"
          >
            Lists
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      LISTS
    </h3>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td>
            Name
          </td>
          <td>
            # Items
          </td>
          <td>
            <button
              class="button is-small"
              data-testid="addNew"
            >
              <span
                class="icon is-small"
              >
                <i
                  class="fas fa-plus"
                />
              </span>
            </button>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              class="has-text-black"
              data-testid="viewList1"
              href="/lists/1"
            >
              list1
            </a>
          </td>
          <td>
            <span>
              <center>
                4
              </center>
            </span>
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteList1"
                href="/lists/1/delete"
              />
            </center>
          </td>
        </tr>
        <tr>
          <td>
            <a
              class="has-text-black"
              data-testid="viewList2"
              href="/lists/2"
            >
              list2
            </a>
          </td>
          <td>
            <span>
              <center>
                6
              </center>
            </span>
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteList2"
                href="/lists/2/delete"
              />
            </center>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
